### PR TITLE
Port splitCheckRun of worker, step 1

### PR DESCRIPTION
**Pull request by xiaobogaga**
@coocood PTAL, thanks.

### CODE REVIEW

File: tikv/raftstore/worker.go [hunk 1]
```diff
@@ -180,14 +182,80 @@ func newWorker(name string, wg *sync.WaitGroup) *worker {
...
 func (r *splitCheckRunner) run(t task) {
-	// TODO: stub
+	spCheckTask := t.data.(*splitCheckTask)
+	region := spCheckTask.region
+	regionId := region.Id
+	startKey := EncStartKey(region)
+	endKey := EncEndKey(region)
+	log.Debugf("executing task: [regionId: %d, startKey: %s, endKey: %s]", regionId,
+		hex.EncodeToString(startKey), hex.EncodeToString(endKey))
+	/// todo, a metrics method ignored here.
```
Comment (coocood) [inline tikv/raftstore/worker.go]: We can remove those metrics todos.

File: tikv/raftstore/worker.go [hunk 2]
```diff
@@ -180,14 +182,80 @@ func newWorker(name string, wg *sync.WaitGroup) *worker {
...
+	host := r.coprocessorHost.newSplitCheckerHost(region, r.engine, spCheckTask.autoSplit,
+		spCheckTask.policy)
+	if host.skip() {
+		log.Debugf("skip split check, [regionId : %d]", regionId)
```
Comment (coocood) [inline tikv/raftstore/worker.go]: missing `return`.

File: tikv/raftstore/worker.go [hunk 3]
```diff
@@ -180,14 +182,80 @@ func newWorker(name string, wg *sync.WaitGroup) *worker {
...
+		keys, err = host.approximateSplitKeys(region, r.engine)
+		if err == nil {
+			for i, k := range keys {
+				keys[i] = OriginKey(k)
+			}
+			break
+		}
+		log.Errorf("failed to get approximate split key, try scan way: [regionId: %d, err : %v]",
+			regionId, err)
+		fallthrough
```
Comment (coocood) [inline tikv/raftstore/worker.go]: Use `fallthrough` does avoid some duplicated code.
But we rarely use fallthough, it makes the logic more tangled, harder to read.

File: tikv/raftstore/worker.go [hunk 4]
```diff
@@ -180,14 +182,80 @@ func newWorker(name string, wg *sync.WaitGroup) *worker {
...
+	var keys [][]byte
+	var err error
+	switch host.policy() {
+	case pdpb.CheckPolicy_APPROXIMATE:
+		keys, err = host.approximateSplitKeys(region, r.engine)
+		if err == nil {
+			for i, k := range keys {
+				keys[i] = OriginKey(k)
+			}
+			break
```
Comment (coocood) [inline tikv/raftstore/worker.go]: We rarely use `break` to break `switch` case, use `else` here is better.

File: tikv/raftstore/worker.go [hunk 5]
```diff
@@ -180,14 +182,80 @@ func newWorker(name string, wg *sync.WaitGroup) *worker {
...
+			regionId, err)
+		fallthrough
+	case pdpb.CheckPolicy_SCAN:
+		keys, err = r.scanSplitKeys(host, region, startKey, endKey)
+		if err != nil {
+			log.Errorf("failed to scan split key: [regionId: %d, err: %v]", regionId, err)
+			return
+		}
+	}
+	if keys != nil && len(keys) != 0 {
```
Comment (coocood) [inline tikv/raftstore/worker.go]: It's safe to call `len(keys)` even if keys is nil.

File: tikv/raftstore/coprocessor.go [hunk 6]
```diff
@@ -14,6 +16,48 @@ const (
...
+type splitChecker interface {
+	onKv(_ *observerContext, _ splitCheckKeyEntry) bool
+	splitKeys() [][]byte
+	approximateSplitKeys(_ *metapb.Region, _ *badger.DB) ([][]byte, error)
+	policy() pdpb.CheckPolicy
+}
+
+type splitCheckerHost struct {
+	autoSplit bool
+	checkers  []*splitChecker
```
Comment (coocood) [inline tikv/raftstore/coprocessor.go]: Never use pointer for interface types.

File: tikv/raftstore/coprocessor.go [hunk 7]
```diff
@@ -14,6 +16,48 @@ const (
...
+}
+
+func (spCheckerHost *splitCheckerHost) skip() bool {
+	//Todo, currently it is a place holder
+	return false
+}
+
+func (spCheckerHost *splitCheckerHost) onKv() bool {
```
Comment (coocood) [inline tikv/raftstore/coprocessor.go]: missing arguments.

File: tikv/raftstore/coprocessor.go [hunk 8]
```diff
@@ -14,6 +16,48 @@ const (
 	RegionChangeEvent_Destroy
 )
 
+type observerContext struct {
+	region *metapb.Region
+	bypass bool
+}
+
+type splitChecker interface {
```
Comment (coocood) [inline tikv/raftstore/coprocessor.go]: We also need to port comments.

Comment (xiaobogaga) [reply tikv/raftstore/coprocessor.go]: ok

File: tikv/raftstore/coprocessor.go (hunk 6 above)
Comment (xiaobogaga) [reply tikv/raftstore/coprocessor.go]: sorry

File: tikv/raftstore/coprocessor.go (hunk 7 above)
Comment (xiaobogaga) [reply tikv/raftstore/coprocessor.go]: yes

File: tikv/raftstore/worker.go (hunk 4 above)
Comment (xiaobogaga) [reply tikv/raftstore/worker.go]: yes

File: tikv/raftstore/worker.go (hunk 3 above)
Comment (xiaobogaga) [reply tikv/raftstore/worker.go]: sorry, I would fix this

File: tikv/raftstore/worker.go (hunk 5 above)
Comment (xiaobogaga) [reply tikv/raftstore/worker.go]: ok, thanks

**Comment by coocood** [comment]
LGTM

**Pull request merged**